force-app/main/adapter/in/sobjects/contact/Contacts.cls

```diff
@@ -389,83 +389,104 @@ public inherited sharing class Contacts extends fflib_SObjects {
-                mapConAddr.put(con, npspAddressFromContact.getRecord());
-            }
+    public void createOrUpdateContactAddress(List<Contact> contacts) {
+        if (contacts.isEmpty()) {
+            return;
+        }
+        // Strategy:
+        // create a new (unsaved) address for each contact.
+        // see if there is an existing address that matches each new address.
+        //
+        // for existing addresses we match
+        //     process during BeforeInsert/Update
+        //     update appropriate fields on contact
+        //
+        // for new addresses that won't be a default (ie, the contact has override set)
+        //     process during BeforeInsert/Update
+        //     create the address now
+        //     update appropriate fields on contact
+        //
+        // for new addresses that will be a default address
+        //     process during AfterInsert/Update
+        //     mark the new address as default
+        //     put it on dmlWrapper
+        //
 
-            // look for duplicates for our proposed new addresses
-            Map<Address__c, Address__c> mapAddrDupAddr = Addresses.getExistingAddresses(mapConAddr
-                    .values());
+        Map<Contact, Address__c> contactAddressesByContact = getContactAddressesByContact(contacts);
 
-            // maps & lists to keep track of the addresses we must insert or update
-            Map<Contact, Address__c> mapConAddrInsertNow = new Map<Contact, Address__c>();
+        // look for duplicates for our proposed new addresses
+        Map<Address__c, Address__c> existingAddressesByAddress =
+                Addresses.getExistingAddresses(contactAddressesByContact.values());
 
-            for (Contact con : listCon) {
-                Address__c addrNew = mapConAddr.get(con);
-                Address__c addrDup = mapAddrDupAddr.get(addrNew);
+        // maps & lists to keep track of the addresses we must insert or update
+        Map<Contact, Address__c> contactAddressesToInsertByContact = new Map<Contact, Address__c>();
 
-                // if found a match
-                if (addrDup != null) {
-                    // record the match in the contact
-                    con.Current_Address__c = addrDup.Id;
+        for (Contact contact : contacts) {
+            Address__c newAddressFromContact = contactAddressesByContact.get(contact);
+            Address__c existingAddressFromContact = existingAddressesByAddress.get(newAddressFromContact);
 
-                    //reset the Primary Address Type only; resetting other fields on Contact will create a new Address
-                    con.npe01__Primary_Address_Type__c = con.npe01__Primary_Address_Type__c == null
-                            ? addrDup.Address_Type__c
-                            : con.npe01__Primary_Address_Type__c;
+            // if found a match
+            if (contactAddressHasAddressMatch(existingAddressFromContact)) {
```

Comment:
I do think that `existingAddressFromContact != null` might be ok to have here instead of a function call

Reply:
I think the reason it may be good to keep is that it helps clarify what the null check is doing by wrapping behavior around it. Ideally, this would be a query on an entity to make it a little more straightforward but I think this at least moves us in that direction.